### ISSUE TITLE
Unable to build dovecot from source when --with-mysql added to formula args

### ISSUE THREAD

**Issue by robertmfuhrer** [Homebrew/brew#6719]
# Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

Install dovecot 2.3.8 with MySQL driver support, for userdb and passdb storage, which isn't enabled in the default dovecot formula. (--with-mysql is missing from the formula's configure args.) [As a result of the default config, trying to specify "mysql" in dovecot's auth config produces an error that the mysql driver isn't recognized.]

First, I installed mysql via `brew install mysql`, which succeeded. I was able to run the mysql server, connect to it, and create the necessary tables for dovecot.

Next, I added the --with-mysql arg to the dovecot config via `brew edit dovecot`.

Then, I tried to build+install via "brew install dovecot --build-from-source".

## What happened (include command output)

The build from source fails because the compiler can't find the mysql.h header file.

<details>
  <summary>brew install dovecot --build-from-source</summary>
  <pre>
==> Downloading https://dovecot.org/releases/2.3/dovecot-2.3.8.tar.gz
Already downloaded: /Users/rmfuhrer/Library/Caches/Homebrew/downloads/8f31998f121bbd9d05987ea58227f738137e7de01861fea61a130efd20a88ec8--dovecot-2.3.8.tar.gz
==> ./configure --prefix=/usr/local/Cellar/dovecot/2.3.8 --libexecdir=/usr/local/Cellar/dovecot/2.3.8/libexec --
==> make install
Last 15 lines from /Users/rmfuhrer/Library/Logs/Homebrew/dovecot/02.make:
#include <mysql.h>
         ^~~~~~~~~
1 error generated.
  </pre>

Here's the relevant excerpt from the make logs:

  <pre>
  libtool: compile: clang -DHAVE_CONFIG_H -I. -I../.. -I../../src/lib -I../../src/lib-settings -I/usr/local/include/mysql -std=gnu99 -g -O2 -fstack-protector-strong -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -Wall -W -Wmissing-prototypes -Wmissing-declarations -Wpointer-arith -Wchar-subscripts -Wformat=2 -Wbad-function-cast -Wno-duplicate-decl-specifier -Wstrict-aliasing=2 -c driver-mysql.c -fno-common -DPIC -o .libs/driver-mysql.o

driver-mysql.c:20:10: fatal error: 'mysql.h' file not found
#include <mysql.h>
         ^~~~~~~~~
  </pre>
</details>

## What you expected to happen

Notice that the clang command above specifies -I/usr/local/include/mysql, which is the correct directory containing the mysql.h header file, which does in fact exist. That entire directory and its contents are world-readable:

  <pre>
$ ls -l -d /usr/local/include/mysql
lrwxr-xr-x  1 rmfuhrer  admin  36 Nov 10 12:33 /usr/local/include/mysql@ -> ../Cellar/mysql/8.0.18/include/mysql
$ ls -l -d /usr/local/Cellar/mysql/8.0.18/include
drwxr-xr-x  3 rmfuhrer  staff  96 Sep 20 04:59 /usr/local/Cellar/mysql/8.0.18/include/
$ ls -l -d /usr/local/Cellar/mysql/8.0.18/include/mysql/
drwxr-xr-x  16 rmfuhrer  staff  512 Sep 20 04:59 /usr/local/Cellar/mysql/8.0.18/include/mysql//
$ ls -l -d /usr/local/Cellar/mysql/8.0.18/include/mysql/mysql.h 
-rw-r--r--  1 rmfuhrer  staff  32452 Sep 20 04:59 /usr/local/Cellar/mysql/8.0.18/include/mysql/mysql.h
$
  </pre>

If I run the exact compile command cited in the make logs directly from bash, on a tiny file that has just the #include statement, it runs correctly:

 <pre>
$ clang -DHAVE_CONFIG_H -I. -I../.. -I../../src/lib -I../../src/lib-settings -I/usr/local/include/mysql -std=gnu99 -g -O2 -fstack-protector-strong -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -Wall -W -Wmissing-prototypes -Wmissing-declarations -Wpointer-arith -Wchar-subscripts -Wformat=2 -Wbad-function-cast -Wno-duplicate-decl-specifier -Wstrict-aliasing=2 -c driver-mysql.c  -fno-common -DPIC -o /tmp/driver-mysql.o
$
 </pre>

That said, I realize that brew has its own wrappers for most of the tool chain (compilers, linkers, etc), and the environment vars are probably also different.

I tried uninstalling both mysql and dovecot, and trying again, but got the same result.

FWIW, "brew install dovecot" (i.e. not building from source, and hence without the --with-mysql arg) works fine, but then of course dovecot doesn't recognize the MySQL driver when specified in the auth configs.

Output from "brew config" and "brew doctor".
<pre>
$ brew config
HOMEBREW_VERSION: 2.1.16-38-ge2c76cc
ORIGIN: https://github.com/Homebrew/brew
HEAD: e2c76cce8e01fd80e0910ddbf8d03092d8753673
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a5561705c81aa5279992353fe1a9836acc59d53d
Core tap last commit: 11 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_VISUAL: vi
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: 10.0 build 1001
Git: 2.20.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 12
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 10.3
CLT headers: 10.3.0.0.1.1562985497
                       [[ rmfuhrer@BBs-Mac-mini 12:42:17 /usr/local/etc/dovecot ]]
$ brew doctor
Your system is ready to brew.
$
</pre>

## Step-by-step reproduction instructions (by running `brew` commands)

$ brew update
$ brew install mysql
$ brew edit dovecot
$ brew install dovecot --build-from-source
$

gist-logs from the install attempt:

  https://gist.github.com/8d80739dcfd3225b58d6266ab2edfc17

**Comment by robertmfuhrer** [comment]
I neglected to provide details on my MySQL install (also via brew):

<pre>
$ brew info mysql
mysql: stable 8.0.18 (bottled)
Open source relational database management system
https://dev.mysql.com/doc/refman/8.0/en/
Conflicts with:
  mariadb (because mysql, mariadb, and percona install the same binaries.)
  mariadb-connector-c (because both install plugins)
  mysql-connector-c (because both install MySQL client libraries)
  percona-server (because mysql, mariadb, and percona install the same binaries.)
/usr/local/Cellar/mysql/8.0.18 (293 files, 280.9MB) *
  Poured from bottle on 2019-11-10 at 12:33:35
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/mysql.rb
==> Dependencies
Build: cmake ✔
Required: openssl@1.1 ✔
==> Requirements
Required: macOS >= 10.10 ✔
==> Caveats
We've installed your MySQL database without a root password. To secure it run:
    mysql_secure_installation

MySQL is configured to only allow connections from localhost by default

To connect run:
    mysql -uroot

To have launchd start mysql now and restart at login:
  brew services start mysql
Or, if you don't want/need a background service you can just run:
  mysql.server start
==> Analytics
install: 77,262 (30 days), 212,128 (90 days), 814,985 (365 days)
install_on_request: 72,529 (30 days), 198,968 (90 days), 759,610 (365 days)
build_error: 0 (30 days)
$
</pre>

**Comment by MikeMcQuaid** [comment]
Sorry, we do not support errors when building from source. If you want MySQL support for this formula I suggest you use `brew extract` and follow https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap to maintain your own tap.

**Comment by robertmfuhrer** [comment]
I understand that people can do arbitrary things when building from source that have nothing to do with brew, but AFAICT, the problem stems from the tool wrappers that brew supplies, since the exact same compile command invoked manually (outside brew) works fine from a normal bash shell.

**Comment by jonchang** [comment]
> since the exact same compile command invoked manually (outside brew) works fine from a normal bash shell

This is not a bug and is intentional behavior; see https://docs.brew.sh/Building-Against-Non-Homebrew-Dependencies for details. Please use the Discourse forum in the future if you need help creating formulae in Homebrew.

**Comment by robertmfuhrer** [comment]
I'm confused by your answer -- I wasn't attempting to build against any non-homebrew dependencies. I installed MySQL via "brew install mysql", which includes the necessary libraries and headers.

**Comment by MikeMcQuaid** [comment]
@robertmfuhrer Homebrew is designed to filter out things that aren't declared as dependencies. We do not wish to support either problems that arise when building from source or modified versions of formulae that have different dependencies. You will have to figure that out and handle it yourself: we do not have the resources or inclination to help you here, sorry.

**Comment by robertmfuhrer** [comment]
I know this bug is closed, but I thought I'd close the loop, in case anyone else came looking here and was interested in knowing the outcome.

The answer turned out to be amazingly simple, I'm not sure why no one suggested it (though responses sorted of hinted in the vague direction). If I had been just a little more experienced with brew, the answer probably would have been obvious to me much earlier.

I needed to add this to the formula:

  depends_on "mysql"

Part of what threw me off was people saying "we don't support non-brew dependencies", but this isn't a non-brew dependency, since I installed mysql via brew.

All the same, brew builds filter out non-declared header and library dependencies, so even though the "clang" compile command looked correct, that was the wrapper invocation, and the -I/usr/local/include/mysql was getting stripped out.

I wish the logs would say something when explicit flags to wrapped build tools get stripped out.

That would have made the problem and its solution blindingly obvious at first glance.
